keps/sig-node/20200203-memory-manager.md

```diff
@@ -0,0 +1,741 @@
+Once *Memory Manager* completes the computations, it returns the hints to *Topology Manager* so that *Topology Manager* can figure out which NUMA node or a group of NUMA nodes are the best fit for memory pinning, for a container. The overall calculation is performed for all containers in the pod, and if none of containers is rejected, the pod becomes finally admitted and deployed.
+
+In the deployment phase, *Memory Manager* updates its `Node Map` to reflect the amount of memory and hugepages requested by a container. After that, *Memory Manager* enforces the consumption of memory and hugepages for the container. The enforcement limits the container's memory consumption to the NUMA node or the group of NUMA nodes indicated by *Topology Manager*. In the final step in figure above, *Memory Manager* updates `cgroups`, namely `cpuset.mems` (more details in section [CPUSET enforcement](#cpuset-enforcement)). 
+
+
+Nonetheless, the management of memory is sustained by means of various mechanisms independent of *Memory Manager*. In the situation when memory limits are exceeded (e.g. by containers), a K8S node can trigger one of the mechanisms described in [this](#kubernetes-nodes-memory-management-mechanisms-and-their-relation-to-the-memory-manager) section. The section discusses how those mechanisms relate to *Memory Manager*. At the moment, let us focus on the machansim relevant for the admission of pods in Guaranteed QoS class. The mechanism involves Out-of-Memory (OOM) kernel killer. Operating system triggers OOM killer and kills processes with the accordance to OOM score ranking (more details in section [OOM score enforcement](#oom-score-enforcement)). Low-priority containers (BestEffort/QoS or Burstable/QoS) are targeted before the high-priority containers (Guaranteed/QoS). Therefore, *Memory Manager* will offer guaranteed memory allocation (only) for pods in Guaranteed QoS class. 
+
+<!-- Note that containers of Guaranteed Pod of which Memory Manager takes care are guaranteed to consume allocated memory and hugepages on the designated NUMA node while other Pods are deployed later, by utilizing [OOM score enforcement](#oom-score-enforcement) system of Kubelet. -->
+
+![alt text](./multi_numa.png "Multi-NUMA Partitioning")
+
+*Multi-NUMA* feature is enabled by treating a group of NUMA nodes as a unit, as illustrated in the figure above. The groups can be statically configured with a new _kubelet_ `--multi-numa-groups` flag (more details in section [Multi-NUMA Groups Flag](#multi-numa-groups-flag)). The groups must be disjoint, however, as shown in the figure above. For instance, `[0]`, `[1, 2]`, `[3]` groups are disjoint. This is because the groups that overlap do not ensure guaranteed memory allocation across multiple NUMA nodes (e.g. `[0, 1]`, `[1, 2]`, `[3]`). In other words, if a pair of groups overlap, memory resources of some group can be prior consumed by another group so that the other group will starve.
```

Comment:
Done:
> Figure above illustrates three groups of NUMA nodes. The groups can be statically configured with a new _kubelet_ `--numa-groups` flag (more details in section [Multi-NUMA Groups Flag](#multi-numa-groups-flag)). *Multi-NUMA* strategy treats a group of NUMA nodes as a unit, and it will combine the memory from the group.

The details were moved to the "(#multi-numa-groups-flag)" section.